sw/host/opentitanlib/src/transport/ioexpander/sx1503.rs

```diff
@@ -0,0 +1,240 @@
+        self.set(Some(mode), None, None, None)
+    }
+
+    fn set_pull_mode(&self, mode: PullMode) -> Result<()> {
+        self.set(None, None, Some(mode), None)
+    }
+
+    fn set(
+        &self,
+        mode: Option<PinMode>,
+        value: Option<bool>,
+        pull: Option<PullMode>,
+        analog_value: Option<f32>,
+    ) -> Result<()> {
+        if let Some(_) = analog_value {
+            bail!(TransportError::UnsupportedOperation);
+        }
+        if let (None, None, None) = (mode, value, pull) {
+            return Ok(());
+        }
+        self.sx1503.apply_mux_strapping()?;
+        if let Some(value) = value {
+            self.sx1503.set_or_clear_bit(
+                if self.pin_no < 8 {
+                    Sx1503Registers::DataA
+                } else {
+                    Sx1503Registers::DataB
+                },
+                self.pin_no & 0x07,
+                value,
+            )?;
+        }
+        if let Some(pull) = pull {
+            let (up, down) = match pull {
+                PullMode::None => (false, false),
+                PullMode::PullUp => (true, false),
+                PullMode::PullDown => (false, true),
+            };
+            self.sx1503.set_or_clear_bit(
+                if self.pin_no < 8 {
+                    Sx1503Registers::PullUpA
+                } else {
+                    Sx1503Registers::PullUpB
+                },
+                self.pin_no & 0x07,
+                up,
+            )?;
+            self.sx1503.set_or_clear_bit(
+                if self.pin_no < 8 {
+                    Sx1503Registers::PullDownA
+                } else {
+                    Sx1503Registers::PullDownB
+                },
+                self.pin_no & 0x07,
+                down,
+            )?;
+        }
+        if let Some(mode) = mode {
+            let direction = match mode {
+                PinMode::Input => true,
+                PinMode::PushPull => false,
+                _ => return Err(GpioError::UnsupportedPinMode(mode).into()),
+            };
+            self.sx1503.set_or_clear_bit(
+                if self.pin_no < 8 {
+                    Sx1503Registers::DirA
+                } else {
+                    Sx1503Registers::DirB
+                },
+                self.pin_no & 0x07,
+                direction,
+            )?;
+        }
+        self.sx1503.remove_mux_strapping()?;
+        Ok(())
+    }
+}
+
+/// Creates an instance of `IoExpander` to represent a SX1503 chip as specified in the given
+/// configuration declaration section.
+pub fn create(
+    conf: &config::IoExpander,
+    transport_wrapper: &TransportWrapper,
+) -> Result<IoExpander> {
+    // Look up I2C bus and strappings, create Sx1503 struct.
+    let Some(ref i2c_bus) = conf.i2c_bus else {
+        bail!("Missing i2c bus numer");
+    };
+    let Some(i2c_addr) = conf.i2c_address else {
+        bail!("Missing i2c addressr");
```

Comment:
These `bail!` statements really should be returning a structured error.  Ok to send in a follow-up PR.

Reply:
Yes.  I have also been thinking several times, that error messages arising from configuration files should refer to the line number(s) and file name(s) in question.  (The "inconsistent configuration" is a notable example, which really should carry two references, to make it clear to the reader what the problem is.)  I will start looking into how to achieve that.